internal/validation/validator.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package validation
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/go-playground/validator/v10"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+	"github.com/canonical/identity-platform-admin-ui/internal/tracing"
+)
+
+const apiVersion = "v0"
+
+type ValidatingFunc func(r *http.Request) validator.ValidationErrors
+
+type Validator struct {
+	validatingFuncs map[string]ValidatingFunc
+
+	tracer  tracing.TracingInterface
+	monitor monitoring.MonitorInterface
+	logger  logging.LoggerInterface
+}
+
+func (v *Validator) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx, span := v.tracer.Start(r.Context(), "validator.Validator.Middleware")
+		r = r.WithContext(ctx)
+		defer span.End()
+
+		prefix := v.prefix(r.URL.Path)
+		vf, isRegistered := v.validatingFuncs[prefix]
+
+		if isRegistered {
+			if validationErr := vf(r); validationErr != nil {
+				e := NewValidationError(validationErr)
+
+				w.WriteHeader(e.Status)
+				_ = json.NewEncoder(w).Encode(e)
+				return
+			}
+		}
+
+		next.ServeHTTP(w, r)
+	})
+}
+
+func (v *Validator) RegisterValidatingFunc(prefix string, vf ValidatingFunc) error {
+	if vf == nil {
+		return fmt.Errorf("validatingFunc can't be null")
+	}
+
+	if _, isRegistered := v.validatingFuncs[prefix]; isRegistered {
+		return fmt.Errorf("prefix is already registered")
+	}
+
+	v.validatingFuncs[prefix] = vf
+
+	return nil
+}
+
+func (v *Validator) prefix(path string) string {
+	after, found := strings.CutPrefix(path, fmt.Sprintf("/api/%s/", apiVersion))
```

Comment:
as discussed, let's have a look again at the naming here, `path` doesn't sound correct